pkgs/applications/virtualization/gvisor/default.nix

```diff
@@ -0,0 +1,176 @@
+{
+  stdenv
+, pkgs
+, fetchFromGitHub
+, cacert
+, symlinks
+, writeScript
+
+, coreutils
+, bash
+, bazel
+, git
+, go
+, python
+}:
+
+let
+
+  # gvisor doesn't have any releases yet
+  version = "2018-11-10";
+
+  # From the Bazel docs:
+  #   Beneath the outputUserRoot directory, we also create an outputBase
+  #   directory whose name is the MD5 hash of the path name of the workspace
+  #   directory.
+  #
+  # The default `fetchzip` arguments will unpack our file as the name `source`,
+  # and the Nix sandbox builds things in the `/build` directory. Thus, we can
+  # predict this hash and calculate it here:
+  outputBaseHash = builtins.hashString "md5" "/build/source";
+
+  # Source from GitHub with additional patches.
+  patchedSource = fetchFromGitHub {
+    owner = "google";
+    repo  = "gvisor";
+    rev   = "d97ccfa346d23d99dcbe634a10fa5d81b089100d";
+
+    # NOTE: this is the output of the whole fixed-output derivation, so
+    # `nix-prefetch-git` won't work to obtain this. The easiest way is to just
+    # change it and see what breaks :)
+    sha256 = "1bcnq7kazbf6l5j0g82x2lvg1nbp7z70klk139dxi0jkw0j8dh3r";
+
+    # Patch the source to:
+    #   - Use our host Go toolchain
+    #   - Fix hard-coded paths to Bash, coreutils, etc.
+    #   - Fetch a specific version of abseil instead of the `master` branch
+    extraPostFetch =
+      let
+        # master as of 2018-11-10
+        abseilRev  = "070f6e47b33a2909d039e620c873204f78809492";
+        abseilHash = "a4298dca4149157b379588bec19493bcab56f8b3fb119ea81303b04d70af1b48";
+
+      in ''
+        substituteInPlace "$out/WORKSPACE" \
+          --replace 'go_register_toolchains(go_version="1.11.2")' 'go_register_toolchains(go_version="host")' \
+          --replace \
+            'urls = ["https://github.com/abseil/abseil-cpp/archive/master.zip"],' \
+            'urls = ["https://github.com/abseil/abseil-cpp/archive/${abseilRev}.zip"], sha256 = "${abseilHash}",'
+
+        find "$out" -name '*.sh' -exec \
+          sed -i 's|#!/bin/bash|#!${bash}/bin/bash|g' {} \;
+
+        find "$out" -name '*_test.go' -exec \
+          sed -i \
+            -e 's|/bin/sleep|${coreutils}/bin/sleep|g' \
+            -e 's|/bin/true|${coreutils}/bin/true|g' \
+            {} \;
+      '';
+  };
+
+  # Bazel command we run.
+  bazelCmd = "USER=nix bazel";
+
+  # Use Bazel to fetch dependencies as a fixed-output derivation, so that we
+  # have network access.
+  bazelDependencies = stdenv.mkDerivation rec {
+    inherit version;
+    name = "gvisor-build-dependencies-${version}";
+
+    # The actual source
+    src = patchedSource;
+
+    outputHashMode = "recursive";
+    outputHashAlgo = "sha256";
+    outputHash = "0430pn3q71r6pyxq32k2n1zhnp9hvs5mizvw3zy6zwrsv3fchdb6";
```

Comment:
This hash is also not the expected one when I locally build it. But this could be related to the update of the hash of `patchedSource`.

Reply:
I force-pushed the fix for the comment above. If this hash still doesn't work, mind running this command for me (with the correct store path) and uploading the results to Gist / Pastebin / something?

```
(cd /nix/store/35j7izc656kyppz5nqci9c6rivp2zi9s-gvisor-build-dependencies-2018-11-10 && find . -type f | xargs shasum | sort -k 2,2)
```

Reply:
@andrew-d Hashes are not corrects:(
Let me know if https://gist.github.com/nlewo/bbd43f6a7c985e6d70402ac55a439116 helps you: there are hashes of the resulting build temporary directory (`nix-store -rK ...`).

Reply:
Okay, this is going to be _substantially_ more annoying than I'd expected.  After a bunch of digging, here's what I've found:

- `rules_go` uses some helper tools, which they [`go install`](https://github.com/bazelbuild/bazel-gazelle/blob/dfcaf2121aaa266dfe16646767b8ef805603898a/internal/go_repository.bzl#L240) into a synthetic repository
- Bazel creates `.marker` files to track whether repositories are up-to-date (and which is the vast majority of what differs between our two systems); these marker files appear to include the hash of the underlying files in the working tree.
- Since these are built using the regular `go` tooling, and inconsistent paths, they don't have a consistent output hash.
- Bazel will verify marker as part of the build process, so we can't patch these tools post-hoc (since this means the hashes in the marker files don't match and Bazel tries to re-download everything).

I'm honestly at a bit of a loss; here are my thoughts:
1. Try to get these tools building in a reproducible fashion (requires an upstream patch in `rules_go`)
2. Do something to fix these specific files; the inconsistency comes from a specific debug section in the output binaries (`.note.go.buildid`), so we could try to zero out that section.

Of the two, I'm going to try to do #1, since the second feels fragile to me. But overall, yeah, this is pretty annoying :unamused: 

(also, I thought about trying to build things with Nix itself, but unlike the *tonix utilities that other languages' package managers use, Bazel repository rules allow running arbitrary shell scripts, so I think we'll always have to run Bazel itself to fetch dependencies)